File: DSA-Python/5-array/18-max-subarray-sum.py
# Given an integer array nums, find the subarray with the largest sum, and return its sum.



# brute force - O(n²), O(1)
def max_subarray_sum_1(nums:list[int]) -> int:
    n , max_sum = len(nums), float('-inf')
    
    for i in range(n):
        sum = 0
        
        for j in range(i, n):
            sum += nums[j]
            
            if sum > max_sum:
                max_sum = sum
                
    return max_sum



# kadane - O(n), O(1)
def max_subarray_sum_2(nums:list[int]) -> int:
    n, max_sum, sum = len(nums), float('-inf'), 0
    
    for x in nums:
        sum += x
        
        if sum > max_sum:
            max_sum = sum
            
        if sum < 0:
            sum = 0
            
    return max_sum



if __name__ == "__main__":
    loop = int(input())
    
    for _ in range(loop):
        nums = list(map(int, input().split()))
        
        res = max_subarray_sum_1(nums)
        
        res = max_subarray_sum_2(nums)
        
        print(res)
        

File: DSA-C++/6-array/18-max-subarray-sum.cpp
#include <bits/stdc++.h>

using namespace std;
// Given an integer array nums, find the subarray with the largest sum and return the sum of the elements present in that subarray. A subarray is a contiguous non-empty sequence of elements within an array.




// Brute force - O(n²), O(1)
int max_subarray_sum_1(vector <int> &nums) {
    int n = nums.size(), max_sum=INT_MIN;

    for (int i=0; i<n; ++i) {
        int sum =0;

        for (int j=i; j<n; ++j) {
            sum += nums[j];

            if (sum > max_sum) {
                max_sum = sum;
            }
        }
    }

    return max_sum;
}


// Kadane algorithm - O(n), O(1)
int max_subarray_sum_2(vector <int> &nums) {
    int max_sum = INT_MIN, sum = 0;
    
    for (int x:nums) {
        sum += x;

        if (sum > max_sum) {
            max_sum = sum;
        }

        if (sum < 0) {
            sum = 0;
        }
    }

    return max_sum;
}

int main() {
    vector <int> nums = {2, 3, 5, -2, 7, -4};
    nums = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    nums = {-3, -2, -5};
    // nums = {-2, 0, -1};
    // nums = {};

    cout << max_subarray_sum_1(nums) << endl;

    cout << max_subarray_sum_2(nums) << endl;

    return 0;
}